src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
+		if (dram_addr < RAM_SIZE)
+			ram[dram_addr] = static_cast<uint8_t>(val);
 		break;
 	default:
 #if LOG_GUS
-		LOG_MSG("Write GUS at port 0x%x with %x", port, val);
+		LOG_MSG("GUS: Write to port 0x%x with value %x", port, val);
 #endif
 		break;
 	}
 }
 
-static void GUS_DMA_Callback(DmaChannel * chan,DMAEvent event) {
-	if (event!=DMA_UNMASKED) return;
-	Bitu dmaaddr;
-	//Calculate the dma address
-	//DMA transfers can't cross 256k boundaries, so you should be safe to just determine the start once and go from there
-	//Bit 2 - 0 = if DMA channel is an 8 bit channel(0 - 3).
-	if (myGUS.DMAControl & 0x4)
-		dmaaddr = (((myGUS.dmaAddr & 0x1fff) << 1) | (myGUS.dmaAddr & 0xc000)) << 4;
-	else
-		dmaaddr = myGUS.dmaAddr << 4;
-	//Reading from dma?
-	if((myGUS.DMAControl & 0x2) == 0) {
-		Bitu read=chan->Read(chan->currcnt+1,&GUSRam[dmaaddr]);
-		//Check for 16 or 8bit channel
-		read*=(chan->DMA16+1);
-		if((myGUS.DMAControl & 0x80) != 0) {
-			//Invert the MSB to convert twos compliment form
-			Bitu i;
-			if((myGUS.DMAControl & 0x40) == 0) {
-				// 8-bit data
-				for(i=dmaaddr;i<(dmaaddr+read);i++) GUSRam[i] ^= 0x80;
-			} else {
-				// 16-bit data
-				for(i=dmaaddr+1;i<(dmaaddr+read);i+=2) GUSRam[i] ^= 0x80;
-			}
-		}
-	//Writing to dma
-	} else {
-		chan->Write(chan->currcnt+1,&GUSRam[dmaaddr]);
-	}
-	/* Raise the TC irq if needed */
-	if((myGUS.DMAControl & 0x20) != 0) {
-		myGUS.IRQStatus |= 0x80;
-		GUS_CheckIRQ();
+void Gus::UpdatePeakAmplitudes(const float *stream)
+{
+	for (int i = 0; i < BUFFER_SAMPLES - 1; i += 2) {
+		peak.left = std::max(peak.left, fabsf(stream[i]));
+		peak.right = std::max(peak.right, fabsf(stream[i + 1]));
 	}
-	chan->Register_Callback(0);
 }
 
-static void GUS_CallBack(uint16_t len)
+void Gus::UpdateWaveLsw(int32_t &addr) const
 {
-	Bit32s buffer[MIXER_BUFSIZE][2];
-	memset(buffer, 0, len * sizeof(buffer[0]));
-
-	for (Bitu i = 0; i < myGUS.ActiveChannels; i++) {
-		guschan[i]->generateSamples(buffer[0], len);
-	}
-	for (Bitu i = 0; i < len; i++) {
-		buffer[i][0] >>= VOL_SHIFT;
-		buffer[i][1] >>= VOL_SHIFT;
-	}
-	gus_chan->AddSamples_s32(len, buffer[0]);
-	CheckVoiceIrq();
+	const auto lower = static_cast<unsigned>(addr) & WAVE_LSW_MASK;
+	addr = static_cast<int32_t>(lower | register_data);
 }
 
-// Generate logarithmic to linear volume conversion tables
-static void MakeTables(void) {
-	int i;
-	double out = (double)(1 << 13);
-	for (i=4095;i>=0;i--) {
-		vol16bit[i]=(Bit16s)out;
-		out/=1.002709201;		/* 0.0235 dB Steps */
-		//Original amplification routine in the hardware
-		//vol16bit[i] = ((256 + i & 0xff) << VOL_SHIFT) / (1 << (24 - (i >> 8)));
-	}
-	pantable[0] = 4095 << RAMP_FRACT;
-	for (i=1;i<16;i++) {
-		pantable[i]=(Bit32u)(0.5-128.0*(log((double)i/15.0)/log(2.0))*(double)(1 << RAMP_FRACT));
-	}
+void Gus::UpdateWaveMsw(int32_t &addr) const
+{
+	const uint32_t upper = register_data & 0x1fff;
+	const auto lower = static_cast<unsigned>(addr) & WAVE_MSW_MASK;
```

Comment:
`static_cast<unsigned>` does not change `int32_t` into an `uint32_t`. It changes `int32_t` into `unisgned int`.

Do we really need WAVE_LSW_MASK and WAVE_MSW_MASK defined top-level? These are being used only in UpdateWaveLsw and UpdateWaveMsw - it would be easier to read if they were simply defined in the functions as bitmasks (like in line 1253).

*Note for myself: double-check if lines 1254 and 1255 are actually correct; they seem weird.*

Reply:
> Do we really need WAVE_LSW_MASK and WAVE_MSW_MASK defined top-level

Good catch - now defined locally.